### ISSUE TITLE
Deriving Deserialize for structs with Cow<'a, str> fails

### ISSUE THREAD

**Issue by vks** [serde-rs/serde#507]
Consider the following code:
```rust
#[derive(Deserialize)]
struct S<'a>(Cow<'a, str>);
```
This fails with a cryptic error message:
```
error[E0207]: the lifetime parameter `'a` is not constrained by the impl trait, self type, or predicates
 --> <quote expansion>:1:2
  |
1 | <'a>
  |  ^^ unconstrained lifetime parameter
```
I believe `Deserialize` is implemented for `Cow<'a, str>`, so I think this should work.

**Comment by dtolnay** [comment]
The buggy generated code looks like this:

```rust
    struct __Visitor;
    impl<'a> _serde::de::Visitor for __Visitor {
//       ^^ unconstrained lifetime parameter
        type Value = S<'a>;
```

**Comment by oli-obk** [comment]
we can assume that any lifetimes during deserialization are `'static` and thus simply replace then. They should be coercible to the target lifetime. If/When we get a deserialization strategy that borrows the object that it deserializes from, we can revise this.

**Comment by dtolnay** [comment]
Interestingly this works and I can deserialize `S<str>`:

```rust
#[derive(Deserialize)]
#[serde(bound = "T::Owned: Deserialize")]
struct S<'a, T: ?Sized + 'a + ToOwned>(Cow<'a, T>);
```

because it generates this code:

```rust
struct __Visitor<'a, T: ?Sized + 'a + ToOwned>(::std::marker::PhantomData<&'a ()>,
                                               ::std::marker::PhantomData<T>)
    where T::Owned: Deserialize;
impl<'a, T: ?Sized + 'a + ToOwned> _serde::de::Visitor for __Visitor<'a, T>
    where T::Owned: Deserialize
{
    type Value = S<'a, T>;
```

@oli-obk maybe we should stick with this `PhantomData<&'a ()>` approach and just make it work when there are no type parameters.

**Comment by oli-obk** [comment]
> Interestingly this works

amusing.

> maybe we should stick with this PhantomData<&'a ()> approach and just make it work when there are no type parameters.

Yea, we probably just need to take out a few things from the bound-attribute code path

**Comment by dtolnay** [comment]
@vks thanks for the report. I released serde 0.8.3 containing a fix.

**Comment by vks** [comment]
Thanks for the quick fix! It works perfectly for me.
